control-plane/agents/ha-node/src/detector.rs

```diff
@@ -0,0 +1,193 @@
+use crate::{
+    path_provider::{CachedNvmePathProvider, NvmePathNameCollection},
+    reporter::PathReporter,
+    Cli,
+};
+use futures::future::FutureExt;
+use nvmeadm::nvmf_subsystem::Subsystem;
+use std::{collections::HashMap, sync::Arc};
+use tokio::time::{sleep, Duration};
+
+/// Possible states of every path record.
+#[derive(Debug, Clone)]
+enum PathState {
+    Good,
+    Suspected,
+    Failed,
+}
+
+/// Object that represents a broken/suspected NVMe path.
+#[derive(Debug, Clone)]
+struct PathRecord {
+    /// Used to allow detection of outdated records.
+    epoch: u64,
+    nqn: String,
+    state: PathState,
+    reporter: Arc<PathReporter>,
+}
+
+impl PathRecord {
+    fn new(nqn: String, epoch: u64, reporter: Arc<PathReporter>) -> Self {
+        Self {
+            nqn,
+            epoch,
+            state: PathState::Good,
+            reporter,
+        }
+    }
+
+    fn get_nqn(&self) -> &str {
+        &self.nqn
+    }
+
+    #[inline]
+    fn get_epoch(&self) -> u64 {
+        self.epoch
+    }
+
+    #[inline]
+    fn set_epoch(&mut self, epoch: u64) {
+        self.epoch = epoch;
+    }
+
+    // Trigger state transition based on 'connecting' state of the underlying NVMe controller.
+    fn report_connecting(&mut self) {
+        match self.state {
+            PathState::Good => {
+                self.state = PathState::Suspected;
+                info!("Target {} transitioned into SUSPECTED state", self.nqn);
```

Comment:
```suggestion
                info!(state=PathState::Suspected, target.nqn=self.nqn, "Target state transition", self.nqn);
```
might need to impl ToString for PathState

Reply:
This version looks nicer, agreed.

Reply:
Done.